pkg/ingester-rf1/metastore/metastore_hack_test.go

```diff
@@ -0,0 +1,253 @@
+		require.NotNil(t, b.marked.Get([]byte(ulid1)))
+		require.Nil(t, b.active.Get([]byte(ulid2)))
+		require.NotNil(t, b.marked.Get([]byte(ulid2)))
+	})
+	// Should move ulid2 from active to marked.
+	_, ok = m.state.active[ulid2]
+	require.False(t, ok)
+	_, ok = m.state.marked[ulid2]
+	require.True(t, ok)
+	// But not affect ulid1.
+	_, ok = m.state.marked[ulid1]
+	require.True(t, ok)
+}
+
+func TestMetastore_doMark(t *testing.T) {
+	c := Config{RetentionPeriod: time.Minute}
+	m, r, err := newMetastoreForTests(t, c)
+	require.NoError(t, err)
+	clock := quartz.NewMock(t)
+	m.clock = clock
+	// Should apply a MarkCommand to the FSM.
+	clock.Advance(time.Second)
+	require.NoError(t, m.doMark())
+	require.Len(t, r.msgs, 1)
+	cmd, ok := r.msgs[0].(*raftlogpb.MarkCommand)
+	require.True(t, ok)
+	require.Equal(t, cmd.Timestamp, uint64(clock.Now().Add(-c.RetentionPeriod).UnixMilli()))
+}
+
+func TestMetastore_applySweep(t *testing.T) {
+	m, _, err := newMetastoreForTests(t, Config{})
+	require.NoError(t, err)
+	clock := quartz.NewMock(t)
+	m.clock = clock
+	// Put two segments in the metastore, 1 minute apart.
+	ulid1 := ulid.MustNew(ulid.Timestamp(clock.Now()), rand.Reader).String()
+	ulid2 := ulid.MustNew(ulid.Timestamp(clock.Now().Add(time.Minute)), rand.Reader).String()
+	withBoltTx(t, m.db, func(t *testing.T, b *metadataBuckets) {
+		require.NoError(t, b.marked.Put([]byte(ulid1), []byte{}))
+		require.NoError(t, b.marked.Put([]byte(ulid2), []byte{}))
+	})
+	m.state.marked[ulid1] = &metastorepb.BlockMeta{Id: ulid1}
+	m.state.marked[ulid2] = &metastorepb.BlockMeta{Id: ulid2}
+	// Should be a no-op, such that no metadata is deleted.
+	_, err = m.state.applySweep(&raftlogpb.SweepCommand{
+		Timestamp: uint64(clock.Now().UnixMilli()),
+	})
+	require.NoError(t, err)
+	withBoltTx(t, m.db, func(t *testing.T, b *metadataBuckets) {
+		require.NotNil(t, b.marked.Get([]byte(ulid1)))
+		require.NotNil(t, b.marked.Get([]byte(ulid2)))
+	})
+	// Should delete ulid1 from the metastore.
+	_, err = m.state.applySweep(&raftlogpb.SweepCommand{
+		Timestamp: uint64(clock.Now().Add(time.Second).UnixMilli()),
+	})
+	require.NoError(t, err)
+	withBoltTx(t, m.db, func(t *testing.T, b *metadataBuckets) {
+		require.Nil(t, b.marked.Get([]byte(ulid1)))
+		require.NotNil(t, b.marked.Get([]byte(ulid2)))
+	})
+	// Should delete ulid1.
+	_, ok := m.state.marked[ulid1]
+	require.True(t, ok)
+	// But not affect ulid2.
+	_, ok = m.state.active[ulid2]
+	require.True(t, ok)
+	// Should delete ulid2 from the metastore.
+	_, err = m.state.applySweep(&raftlogpb.SweepCommand{
+		Timestamp: uint64(clock.Now().Add(time.Minute + time.Second).UnixMilli()),
+	})
+	require.NoError(t, err)
+	withBoltTx(t, m.db, func(t *testing.T, b *metadataBuckets) {
+		require.Nil(t, b.marked.Get([]byte(ulid1)))
+		require.Nil(t, b.marked.Get([]byte(ulid2)))
+	})
+	// Should delete ulid2.
+	_, ok = m.state.marked[ulid2]
+	require.True(t, ok)
+	// ulid1 should still be deleted.
+	_, ok = m.state.marked[ulid1]
```

Comment:
Shouldn't the sweep delete the object from the metastore? Is it expected to still be in the state.marked bucket or should the value be deleted from storage & state?

Reply:
Yes! I copied this tests and forgot to fix them.

Reply:
Yes! I copied these from the other test and forgot to change them.